File: leetcode_java/54. Spiral Matrix.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int h = matrix.length, l = matrix[0].length;
        int i = 0, j = -1;
        while(true){
            for(int k = 0; k < l; k++)
                res.add(matrix[i][++j]);
            if(--h == 0) break;
            for(int k = 0; k < h; k++)
                res.add(matrix[++i][j]);
            if(--l == 0) break;
            for(int k = 0; k < l; k++)
                res.add(matrix[i][--j]);
            if(--h == 0) break;
            for(int k = 0; k < h; k++)
                res.add(matrix[++i][j]);
            if(--l == 0) break;
        }
        return res;
    }
}

File: leetcode_python/54. Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: [[int]]) -> [int]:
        if not matrix: return []
        res = []
        h, l = len(matrix), len(matrix[0])
        i, j = 0, -1
        while True:
            for _ in range(l):
                j += 1
                res.append(matrix[i][j])
            h -= 1
            if not h: break
            for _ in range(h):
                i += 1
                res.append(matrix[i][j])
            l -= 1
            if not l: break
            for _ in range(l):
                j -= 1
                res.append(matrix[i][j])
            h -= 1
            if not h: break
            for _ in range(h):
                i -= 1
                res.append(matrix[i][j])
            l -= 1
            if not l: break
        return res

s = Solution()
s.spiralOrder([[1,2,3],[4,5,6],[7,8,9]])